vhdl_syntax/src/parser/productions/names.rs

```diff
@@ -5,31 +5,506 @@
 
-impl<T: TokenStream> Parser<T> {
-    pub fn designator(&mut self) {
-        self.expect_one_of_tokens([Identifier, StringLiteral, CharacterLiteral]);
+fn is_start_of_attribute_name<T: TokenStream>(parser: &mut Parser<T>) -> bool {
+    // Checking for `LeftSquare || Tick` will result in ambiguities with other grammar rules where a signature is possible right after a name.
+    // Those rules can be `alias_declaration` (LRM §6.6.1) and `subprogram_instantiation_declaration` (LRM §4.4).
+    // By checking whether the closing square bracket is followed by a `Tick` this ambiguity is resolved
+    match parser.peek_token() {
+        Some(Tick) => true,
+        Some(LeftSquare) => {
+            let mut idx = 1;
+            let mut bracket_count = 1;
+
+            while bracket_count > 0 {
+                match parser.peek_nth_token(idx) {
+                    Some(LeftSquare) => bracket_count += 1,
+                    Some(RightSquare) => bracket_count -= 1,
+                    Some(_) => {}
+                    None => {
+                        return false;
+                    }
+                }
+
+                idx += 1;
+            }
+
+            parser.next_nth_is(Tick, idx)
+        }
+        Some(_) | None => false,
     }
+}
 
+impl<T: TokenStream> Parser<T> {
     pub fn name(&mut self) {
+        // (Based on) LRM §8.1
+        // The LRM grammar rules for names were transformed to avoid left recursion.
+
+        // In contrast to the LRM, this parsing routine is greedy. Meaning, it will consume trailing parenthesized
+        // expressions even if the belong to an outer grammar rule!
         self.start_node(Name);
-        // TODO
-        self.designator();
+
+        if self.next_is(LtLt) {
+            self.external_name();
+        } else {
+            self.expect_one_of_tokens([Identifier, StringLiteral, CharacterLiteral]);
+        }
+
+        self.name_tail();
         self.end_node();
     }
 
     pub fn type_mark(&mut self) {
         self.name()
     }
 
-    pub fn opt_label(&mut self) {
+    pub(crate) fn designator(&mut self) {
+        // TODO: That designator is not fully LRM compliant
+        self.expect_one_of_tokens([Identifier, StringLiteral, CharacterLiteral]);
+    }
+
+    pub(crate) fn opt_label(&mut self) {
         if self.next_is(Identifier) && self.next_nth_is(Colon, 1) {
             self.start_node(Label);
             self.skip_n(2);
             self.end_node();
         }
     }
+    pub(crate) fn name_list(&mut self) {
+        self.start_node(NameList);
+        self.separated_list(Parser::name, Comma);
+        self.end_node();
+    }
+
+    fn suffix(&mut self) {
+        // LRM §8.3
+        // suffix ::= identifier | string_literal | character_literal | `all` ;
+        self.expect_one_of_tokens([
+            Identifier,
+            StringLiteral,
+            CharacterLiteral,
+            Keyword(Kw::All),
+        ]);
+    }
+
+    fn name_tail(&mut self) {
```

Comment:
I would probably call this `opt_name_tail`, since this does not fail if there is no name tail present.